File: PROG/Week06/Store.java
import java.util.*;

public class Store {
    private CashRegister cashRegister;
    private LinkedList<Product> products = new LinkedList<Product>();

    public Store() {
        products.add(new Product("Whiteboard Marker", 85, 1.5));
        products.add(new Product("Whiteboard Eraser", 45, 5));
        products.add(new Product("Black Pen", 100, 1.5));
        products.add(new Product("Red Pen", 100, 1.5));
        products.add(new Product("Blue Pen", 100, 1.5));
        cashRegister = new CashRegister();
    }

    public static void main(String[] args) {
        new Store().use();
    }

    private char readChoice() {
        System.out.print("Choice (s/r/v/c/p/x): ");
        return In.nextChar();
    }

    public void use() {
        while (true) {
            switch (readChoice()) {
                case 's':
                    sell();
                    break;
                case 'r':
                    restock();
                    break;
                case 'v':
                    viewStock();
                    break;
                case 'c':
                    viewCash();
                    break;
                case 'p':
                    pruneProducts();
                    break;
                case 'x':
                    System.out.println("Done");
                    System.exit(0);
                    break;
                default:
                    help();
                    break;
            }
        }
    }

    private ArrayList<Product> findProducts(String name) {
        ArrayList<Product> matches = new ArrayList<Product>();
        name = name.toLowerCase();
        for (Product product : products) {
            if (product.getName().toLowerCase().contains(name))
                matches.add(product);
        }
        if (matches.size() == 0)
            return null;
        return matches;
    }

    private void sell() {
        String name = readName();
        ArrayList<Product> ps = findProducts(name);
        if (ps == null) {
            System.out.println("No such product");
            return;
        }
        if (ps.size() > 1) {
            System.out.println("Multiple products match:");
            for (Product product : ps) {
                System.out.println(product);
            }
            return;
        }
        System.out.println("Selling " + ps.get(0).getName());
        int n = readNumber();
        cashRegister.add(ps.get(0).sell(n));
    }

    private void restock() {
        String name = readName();
        ArrayList<Product> ps = findProducts(name);
        if (ps == null) {
            System.out.println("Adding new product");
            int n = readNumber();
            double price = readPrice();
            products.add(new Product(name, n, price));
            return;
        }
        for (Product p : ps) {
            System.out.println("Restocking " + p.getName());
            int n = readNumber();
            p.restock(n);
        }
    }

    private void viewStock() {
        for (Product product : products)
            System.out.println(product);

    }

    private void viewCash() {
        System.out.println(cashRegister);
    }

    private void pruneProducts() {
        ArrayList<Product> toDelete = new ArrayList<>();
        for (Product product : products) {
            if (product.isEmpty())
                toDelete.add(product);
        }
        for (Product product : toDelete) {
            products.remove(product);
        }
    }

    private String readName() {
        System.out.print("Name: ");
        return In.nextLine();

    }

    private double readPrice() {
        System.out.print("Price: $");
        return In.nextDouble();
    }

    private int readNumber() {
        System.out.print("Number: ");
        return In.nextInt();

    }

    private void help() {
        System.out.println("Menu options");
        System.out.println("s = sell");
        System.out.println("r = restock");
        System.out.println("v = view stock");
        System.out.println("c = view cash");
        System.out.println("p = prune products");
        System.out.println("x = exit");
    }
}

File: PROG/Week06/Store.py
from CashRegister import CashRegister
from Product import Product


class Store:
    cashRegister: CashRegister
    products: list[Product] = []

    def __init__(self):
        self.products.append(Product("Whiteboard Marker", 85, 1.5))
        self.products.append(Product("Whiteboard Eraser", 45, 5))
        self.products.append(Product("Black Pen", 100, 1.5))
        self.products.append(Product("Red Pen", 100, 1.5))
        self.products.append(Product("Blue Pen", 100, 1.5))
        self.cashRegister = CashRegister()

    def readChoice(self) -> str:
        return input("Choice (s/r/v/c/p/x): ")[0]

    def use(self):
        while True:
            match self.readChoice():
                case "s":
                    self.sell()
                case "r":
                    self.restock()
                case "v":
                    self.viewStock()
                case "c":
                    self.viewCash()
                case "p":
                    self.pruneProducts()
                case "x":
                    print("Done")
                    exit(0)
                case _:
                    self.help()

    def findProducts(self, name: str) -> list[Product]:
        name = name.lower()
        matches = []
        for product in self.products:
            if name in product.getName().lower():
                matches.append(product)

        if len(matches) == 0:
            return None
        return matches

    def sell(self):
        name = self.readName()
        ps = self.findProducts(name)
        if ps is None:
            print("No such product")
            return

        if len(ps) > 1:
            print("Multiple products match:")
            for product in ps:
                print(product)
            return

        print("Selling " + ps[0].getName())
        n = self.readNumber()
        self.cashRegister.add(ps[0].sell(n))

    def restock(self):
        name = self.readName()
        ps = self.findProducts(name)
        if ps is None:
            print("Adding new product")
            n = self.readNumber()
            price = self.readPrice()
            self.products.append(Product(name, n, price))
            return

        for p in ps:
            print("Restocking " + p.getName())
            n = self.readNumber()
            p.restock(n)

    def viewStock(self):
        for product in self.products:
            print(product)

    def viewCash(self):
        print(self.cashRegister)

    def pruneProducts(self):
        toDelete = []
        for product in self.products:
            if product.isEmpty():
                toDelete.append(product)
        for product in toDelete:
            self.products.remove(product)

    def readName(self) -> str:
        return input("Name: ")

    def readPrice(self) -> float:
        return float(input("Price: $"))

    def readNumber(self) -> int:
        return int(input("Number: "))

    def help(self):
        print("Menu options")
        print("s = sell")
        print("r = restock")
        print("v = view stock")
        print("c = view cash")
        print("p = prune products")
        print("x = exit")


if __name__ == "__main__":
    Store().use()
